// @tanstack-player/plugin-bookmark/dist/index.mjs
// src/bookmark-plugin.ts
var BookmarkEvent = {
  ADDED: "bookmark:added",
  REMOVED: "bookmark:removed"
};
var BookmarkPlugin = class {
  constructor() {
    this.name = "bookmark";
    this.ctx = null;
    this.bookmarks = [];
    this.idCounter = 0;
  }
  setup(ctx) {
    this.ctx = ctx;
  }
  destroy() {
    this.bookmarks = [];
    this.ctx = null;
  }
  /** Add a bookmark at the current playback position (or at a specific time) */
  addBookmark(label, time) {
    const actualTime = time ?? this.ctx?.controller.getCurrentTime() ?? 0;
    const bookmark = {
      id: `bm_${++this.idCounter}_${Date.now()}`,
      time: actualTime,
      label,
      createdAt: Date.now()
    };
    this.bookmarks.push(bookmark);
    this.bookmarks.sort((a, b) => a.time - b.time);
    this.ctx?.eventBus.emit(BookmarkEvent.ADDED, bookmark);
    return bookmark;
  }
  /** Remove a bookmark by ID */
  removeBookmark(id) {
    const index = this.bookmarks.findIndex((b) => b.id === id);
    if (index === -1) return false;
    const [removed] = this.bookmarks.splice(index, 1);
    this.ctx?.eventBus.emit(BookmarkEvent.REMOVED, removed);
    return true;
  }
  /** Get all bookmarks sorted by time */
  getBookmarks() {
    return [...this.bookmarks];
  }
  /** Seek to a specific bookmark */
  seekToBookmark(id) {
    const bookmark = this.bookmarks.find((b) => b.id === id);
    if (!bookmark) return false;
    this.ctx?.controller.seek(bookmark.time);
    return true;
  }
};
function createBookmarkPlugin() {
  return new BookmarkPlugin();
}

export { BookmarkEvent, BookmarkPlugin, createBookmarkPlugin };
