// @tanstack-player/core/dist/index.js
'use strict';

// src/event-bus.ts
var EventBus = class {
  constructor() {
    this.listeners = /* @__PURE__ */ new Map();
  }
  /**
   * Subscribe to an event. Returns an unsubscribe function.
   */
  on(event, listener) {
    if (!this.listeners.has(event)) {
      this.listeners.set(event, /* @__PURE__ */ new Set());
    }
    this.listeners.get(event).add(listener);
    return () => this.off(event, listener);
  }
  /**
   * Subscribe to an event once. Auto-removed after first call.
   */
  once(event, listener) {
    const wrapper = (data) => {
      this.off(event, wrapper);
      listener(data);
    };
    return this.on(event, wrapper);
  }
  /**
   * Unsubscribe a specific listener from an event.
   */
  off(event, listener) {
    const set = this.listeners.get(event);
    if (set) {
      set.delete(listener);
      if (set.size === 0) {
        this.listeners.delete(event);
      }
    }
  }
  /**
   * Emit an event with optional data payload.
   */
  emit(event, data) {
    const set = this.listeners.get(event);
    if (set) {
      for (const listener of [...set]) {
        try {
          listener(data);
        } catch (err) {
          console.error(`[TanStack Player] Error in event listener for "${event}":`, err);
        }
      }
    }
  }
  /**
   * Remove all listeners for all events.
   */
  removeAll() {
    this.listeners.clear();
  }
};

// src/types.ts
var INITIAL_STATE = {
  isPlaying: false,
  isReady: false,
  duration: 0,
  currentTime: 0,
  buffered: 0,
  volume: 1,
  playbackRate: 1,
  isFullscreen: false,
  isMuted: false,
  error: null
};
var PlayerEvent = /* @__PURE__ */ ((PlayerEvent2) => {
  PlayerEvent2["READY"] = "ready";
  PlayerEvent2["PLAY"] = "play";
  PlayerEvent2["PAUSE"] = "pause";
  PlayerEvent2["PROGRESS"] = "progress";
  PlayerEvent2["BUFFER"] = "buffer";
  PlayerEvent2["ENDED"] = "ended";
  PlayerEvent2["ERROR"] = "error";
  PlayerEvent2["SEEK"] = "seek";
  PlayerEvent2["VOLUME_CHANGE"] = "volumechange";
  PlayerEvent2["RATE_CHANGE"] = "ratechange";
  PlayerEvent2["FULLSCREEN_CHANGE"] = "fullscreenchange";
  PlayerEvent2["TIME_UPDATE"] = "timeupdate";
  PlayerEvent2["LOADED_METADATA"] = "loadedmetadata";
  PlayerEvent2["DESTROYED"] = "destroyed";
  return PlayerEvent2;
})(PlayerEvent || {});

// src/state-store.ts
var StateStore = class {
  constructor(initialState) {
    this.listeners = /* @__PURE__ */ new Set();
    this.state = { ...INITIAL_STATE, ...initialState };
  }
  /** Get a snapshot of the current state (immutable) */
  getState() {
    return this.state;
  }
  /** Merge partial updates into state and notify subscribers */
  setState(partial) {
    const prev = this.state;
    this.state = { ...prev, ...partial };
    const changed = Object.keys(partial).some(
      (key) => prev[key] !== this.state[key]
    );
    if (changed) {
      for (const listener of this.listeners) {
        try {
          listener(this.state);
        } catch (err) {
          console.error("[TanStack Player] Error in state listener:", err);
        }
      }
    }
  }
  /** Subscribe to state changes. Returns an unsubscribe function. */
  subscribe(listener) {
    this.listeners.add(listener);
    return () => {
      this.listeners.delete(listener);
    };
  }
  /** Remove all subscribers */
  removeAll() {
    this.listeners.clear();
  }
};

// src/plugin-runtime.ts
var PluginRuntime = class {
  constructor() {
    this.plugins = /* @__PURE__ */ new Map();
  }
  /**
   * Register and setup a plugin.
   */
  register(plugin, controller, eventBus, store) {
    if (this.plugins.has(plugin.name)) {
      console.warn(
        `[TanStack Player] Plugin "${plugin.name}" is already registered. Skipping duplicate.`
      );
      return;
    }
    const ctx = { controller, eventBus, store };
    this.plugins.set(plugin.name, plugin);
    try {
      plugin.setup(ctx);
    } catch (err) {
      console.error(`[TanStack Player] Error setting up plugin "${plugin.name}":`, err);
    }
  }
  /**
   * Get a registered plugin by name.
   */
  get(name) {
    return this.plugins.get(name);
  }
  /**
   * Destroy all registered plugins and clear the registry.
   */
  destroyAll() {
    for (const [name, plugin] of this.plugins) {
      try {
        plugin.destroy?.();
      } catch (err) {
        console.error(`[TanStack Player] Error destroying plugin "${name}":`, err);
      }
    }
    this.plugins.clear();
  }
};

// src/player-controller.ts
var PlayerController = class {
  constructor(options) {
    this.options = options;
    this.destroyed = false;
    this.eventBus = new EventBus();
    this.store = new StateStore();
    this.pluginRuntime = new PluginRuntime();
    this.adapter = options.adapter;
    this.adapter.setEventBridge((event, data) => {
      this.handleAdapterEvent(event, data);
    });
  }
  /**
   * Initialize the player inside a DOM container.
   * Must be called after construction (allows SSR-safe deferred init).
   */
  init(container) {
    const adapterOptions = {
      src: this.options.src,
      autoPlay: this.options.autoPlay,
      controls: this.options.controls,
      fluid: this.options.fluid,
      preload: this.options.preload,
      playbackRates: this.options.playbackRates,
      engineOptions: this.options.engineOptions
    };
    this.adapter.init(container, adapterOptions);
    if (this.options.plugins) {
      for (const plugin of this.options.plugins) {
        this.pluginRuntime.register(plugin, this, this.eventBus, this.store);
      }
    }
  }
  // ─── Playback Controls ─────────────────────────────────────
  play() {
    this.adapter.play();
  }
  pause() {
    this.adapter.pause();
  }
  seek(seconds) {
    this.adapter.seek(seconds);
    this.eventBus.emit("seek" /* SEEK */, { time: seconds });
  }
  setVolume(value) {
    const clamped = Math.max(0, Math.min(1, value));
    this.adapter.setVolume(clamped);
  }
  setMuted(muted) {
    this.adapter.setMuted(muted);
  }
  setPlaybackRate(rate) {
    this.adapter.setPlaybackRate(rate);
  }
  requestFullscreen() {
    this.adapter.requestFullscreen();
  }
  exitFullscreen() {
    this.adapter.exitFullscreen();
  }
  loadSource(src) {
    this.adapter.loadSource(src);
  }
  // ─── Getters ───────────────────────────────────────────────
  getCurrentTime() {
    return this.adapter.getCurrentTime();
  }
  getDuration() {
    return this.adapter.getDuration();
  }
  getVolume() {
    return this.adapter.getVolume();
  }
  getPlaybackRate() {
    return this.adapter.getPlaybackRate();
  }
  isFullscreen() {
    return this.adapter.isFullscreen();
  }
  getState() {
    return this.store.getState();
  }
  getPlugin(name) {
    return this.pluginRuntime.get(name);
  }
  // ─── Lifecycle ─────────────────────────────────────────────
  destroy() {
    if (this.destroyed) return;
    this.destroyed = true;
    this.pluginRuntime.destroyAll();
    this.adapter.destroy();
    this.eventBus.emit("destroyed" /* DESTROYED */);
    this.eventBus.removeAll();
    this.store.removeAll();
  }
  // ─── Internal ──────────────────────────────────────────────
  handleAdapterEvent(event, data) {
    switch (event) {
      case "ready" /* READY */:
        this.store.setState({ isReady: true });
        break;
      case "play" /* PLAY */:
        this.store.setState({ isPlaying: true });
        break;
      case "pause" /* PAUSE */:
        this.store.setState({ isPlaying: false });
        break;
      case "timeupdate" /* TIME_UPDATE */: {
        const payload = data;
        this.store.setState({
          currentTime: payload.currentTime,
          duration: payload.duration
        });
        break;
      }
      case "loadedmetadata" /* LOADED_METADATA */: {
        const payload = data;
        this.store.setState({ duration: payload.duration });
        break;
      }
      case "buffer" /* BUFFER */: {
        const payload = data;
        this.store.setState({ buffered: payload.buffered });
        break;
      }
      case "volumechange" /* VOLUME_CHANGE */: {
        const payload = data;
        this.store.setState({ volume: payload.volume, isMuted: payload.muted });
        break;
      }
      case "ratechange" /* RATE_CHANGE */: {
        const payload = data;
        this.store.setState({ playbackRate: payload.rate });
        break;
      }
      case "fullscreenchange" /* FULLSCREEN_CHANGE */: {
        const payload = data;
        this.store.setState({ isFullscreen: payload.isFullscreen });
        break;
      }
      case "ended" /* ENDED */:
        this.store.setState({ isPlaying: false });
        break;
      case "error" /* ERROR */: {
        const payload = data;
        this.store.setState({ error: payload.error });
        break;
      }
    }
    this.eventBus.emit(event, data);
  }
};

exports.EventBus = EventBus;
exports.INITIAL_STATE = INITIAL_STATE;
exports.PlayerController = PlayerController;
exports.PlayerEvent = PlayerEvent;
exports.PluginRuntime = PluginRuntime;
exports.StateStore = StateStore;
