// @tanstack-player/adapter-videojs/dist/index.mjs
import { PlayerEvent } from '@tanstack-player/core';

// src/videojs-adapter.ts
var VideoJsAdapter = class {
  constructor() {
    this.player = null;
    this.bridge = null;
    this.videoElement = null;
  }
  setEventBridge(bridge) {
    this.bridge = bridge;
  }
  init(container, options) {
    const vjsFn = globalThis.videojs;
    if (!vjsFn) {
      throw new Error(
        "[TanStack Player] Video.js not found. Please load video.js via CDN or import it in your bundle."
      );
    }
    this.videoElement = document.createElement("video");
    this.videoElement.className = "video-js vjs-big-play-centered";
    container.appendChild(this.videoElement);
    const sources = this.normalizeSources(options.src);
    const vjsOptions = {
      controls: options.controls ?? true,
      autoplay: options.autoPlay ?? false,
      preload: options.preload ?? "auto",
      fluid: options.fluid ?? true,
      sources,
      ...options.engineOptions ?? {}
    };
    if (options.playbackRates) {
      vjsOptions.playbackRates = options.playbackRates;
    }
    this.player = vjsFn(this.videoElement, vjsOptions, () => {
      this.emitBridge(PlayerEvent.READY);
    });
    this.bindEvents();
  }
  play() {
    this.player?.play()?.catch?.(() => {
    });
  }
  pause() {
    this.player?.pause();
  }
  seek(seconds) {
    if (this.player) {
      this.player.currentTime(seconds);
    }
  }
  setVolume(value) {
    if (this.player) {
      this.player.volume(value);
    }
  }
  setMuted(muted) {
    if (this.player) {
      this.player.muted(muted);
    }
  }
  setPlaybackRate(rate) {
    if (this.player) {
      this.player.playbackRate(rate);
    }
  }
  requestFullscreen() {
    this.player?.requestFullscreen();
  }
  exitFullscreen() {
    this.player?.exitFullscreen();
  }
  loadSource(src) {
    if (!this.player) return;
    const sources = this.normalizeSources(src);
    this.player.src(sources);
  }
  getCurrentTime() {
    return this.player?.currentTime() ?? 0;
  }
  getDuration() {
    return this.player?.duration() ?? 0;
  }
  getVolume() {
    return this.player?.volume() ?? 1;
  }
  getPlaybackRate() {
    return this.player?.playbackRate() ?? 1;
  }
  isFullscreen() {
    return this.player?.isFullscreen() ?? false;
  }
  destroy() {
    if (this.player) {
      this.player.dispose();
      this.player = null;
    }
    this.videoElement = null;
    this.bridge = null;
  }
  // ─── Internal ───────────────────────────────────────────
  bindEvents() {
    if (!this.player) return;
    this.player.on("play", () => {
      this.emitBridge(PlayerEvent.PLAY);
    });
    this.player.on("pause", () => {
      this.emitBridge(PlayerEvent.PAUSE);
    });
    this.player.on("timeupdate", () => {
      this.emitBridge(PlayerEvent.TIME_UPDATE, {
        currentTime: this.player.currentTime(),
        duration: this.player.duration() || 0
      });
    });
    this.player.on("loadedmetadata", () => {
      this.emitBridge(PlayerEvent.LOADED_METADATA, {
        duration: this.player.duration()
      });
    });
    this.player.on("ended", () => {
      this.emitBridge(PlayerEvent.ENDED);
    });
    this.player.on("volumechange", () => {
      this.emitBridge(PlayerEvent.VOLUME_CHANGE, {
        volume: this.player.volume(),
        muted: this.player.muted()
      });
    });
    this.player.on("ratechange", () => {
      this.emitBridge(PlayerEvent.RATE_CHANGE, {
        rate: this.player.playbackRate()
      });
    });
    this.player.on("fullscreenchange", () => {
      this.emitBridge(PlayerEvent.FULLSCREEN_CHANGE, {
        isFullscreen: this.player.isFullscreen()
      });
    });
    this.player.on("progress", () => {
      const buffered = this.player.bufferedPercent() ?? 0;
      this.emitBridge(PlayerEvent.BUFFER, { buffered });
    });
    this.player.on("error", () => {
      const err = this.player.error();
      this.emitBridge(PlayerEvent.ERROR, {
        error: new Error(err?.message ?? "Unknown player error")
      });
    });
  }
  emitBridge(event, data) {
    this.bridge?.(event, data);
  }
  normalizeSources(src) {
    const sources = Array.isArray(src) ? src : [src];
    return sources.map((s) => ({
      src: s.src,
      type: s.type ?? this.inferType(s.src)
    }));
  }
  inferType(src) {
    if (src.endsWith(".m3u8")) return "application/x-mpegURL";
    if (src.endsWith(".mpd")) return "application/dash+xml";
    if (src.endsWith(".webm")) return "video/webm";
    if (src.endsWith(".ogg") || src.endsWith(".ogv")) return "video/ogg";
    return "video/mp4";
  }
};
function createVideoJsAdapter() {
  return new VideoJsAdapter();
}

export { VideoJsAdapter, createVideoJsAdapter };
