// tanstack/dist/index.cjs
'use strict';

var core = require('@tanstack-player/core');
var adapterVideojs = require('@tanstack-player/adapter-videojs');
var react = require('@tanstack-player/react');
var pluginBookmark = require('@tanstack-player/plugin-bookmark');



Object.defineProperty(exports, "EventBus", {
  enumerable: true,
  get: function () { return core.EventBus; }
});
Object.defineProperty(exports, "INITIAL_STATE", {
  enumerable: true,
  get: function () { return core.INITIAL_STATE; }
});
Object.defineProperty(exports, "PlayerController", {
  enumerable: true,
  get: function () { return core.PlayerController; }
});
Object.defineProperty(exports, "PlayerEvent", {
  enumerable: true,
  get: function () { return core.PlayerEvent; }
});
Object.defineProperty(exports, "PluginRuntime", {
  enumerable: true,
  get: function () { return core.PluginRuntime; }
});
Object.defineProperty(exports, "StateStore", {
  enumerable: true,
  get: function () { return core.StateStore; }
});
Object.defineProperty(exports, "VideoJsAdapter", {
  enumerable: true,
  get: function () { return adapterVideojs.VideoJsAdapter; }
});
Object.defineProperty(exports, "createVideoJsAdapter", {
  enumerable: true,
  get: function () { return adapterVideojs.createVideoJsAdapter; }
});
Object.defineProperty(exports, "TanStackPlayer", {
  enumerable: true,
  get: function () { return react.TanStackPlayer; }
});
Object.defineProperty(exports, "TanStackPlayerContext", {
  enumerable: true,
  get: function () { return react.TanStackPlayerContext; }
});
Object.defineProperty(exports, "useBookmarks", {
  enumerable: true,
  get: function () { return react.useBookmarks; }
});
Object.defineProperty(exports, "usePlayerContext", {
  enumerable: true,
  get: function () { return react.usePlayerContext; }
});
Object.defineProperty(exports, "useProgress", {
  enumerable: true,
  get: function () { return react.useProgress; }
});
Object.defineProperty(exports, "useTanStackPlayer", {
  enumerable: true,
  get: function () { return react.useTanStackPlayer; }
});
Object.defineProperty(exports, "BookmarkEvent", {
  enumerable: true,
  get: function () { return pluginBookmark.BookmarkEvent; }
});
Object.defineProperty(exports, "BookmarkPlugin", {
  enumerable: true,
  get: function () { return pluginBookmark.BookmarkPlugin; }
});
Object.defineProperty(exports, "createBookmarkPlugin", {
  enumerable: true,
  get: function () { return pluginBookmark.createBookmarkPlugin; }
});
