// @tanstack-player/react/dist/index.mjs
import { createContext, useContext, useRef, useState, useEffect, useSyncExternalStore, useCallback, useMemo } from 'react';
import { PlayerController } from '@tanstack-player/core';
export { PlayerEvent } from '@tanstack-player/core';
import { createVideoJsAdapter } from '@tanstack-player/adapter-videojs';
import { jsxs, jsx } from 'react/jsx-runtime';

// src/TanStackPlayer.tsx
var TanStackPlayerContext = createContext(null);
function usePlayerContext() {
  const controller = useContext(TanStackPlayerContext);
  if (!controller) {
    throw new Error(
      "[TanStack Player] useTanStackPlayer must be used inside a <TanStackPlayer /> component."
    );
  }
  return controller;
}
function TanStackPlayer({
  src,
  autoPlay = false,
  controls = true,
  plugins,
  videojsOptions,
  fluid = true,
  preload = "auto",
  playbackRates,
  className,
  style,
  children,
  onReady,
  onPlay,
  onPause,
  onEnded,
  onError,
  onTimeUpdate
}) {
  const containerRef = useRef(null);
  const controllerRef = useRef(null);
  const initializedRef = useRef(false);
  const [controller, setController] = useState(null);
  const normalizedSrc = normalizeSrc(src);
  useEffect(() => {
    if (initializedRef.current || !containerRef.current) return;
    initializedRef.current = true;
    const adapter = createVideoJsAdapter();
    const ctrl = new PlayerController({
      adapter,
      src: normalizedSrc,
      autoPlay,
      controls,
      fluid,
      preload,
      playbackRates,
      engineOptions: videojsOptions,
      plugins
    });
    if (onReady) ctrl.eventBus.on("ready", onReady);
    if (onPlay) ctrl.eventBus.on("play", onPlay);
    if (onPause) ctrl.eventBus.on("pause", onPause);
    if (onEnded) ctrl.eventBus.on("ended", onEnded);
    if (onError) {
      ctrl.eventBus.on("error", (data) => onError(data?.error));
    }
    if (onTimeUpdate) {
      ctrl.eventBus.on("timeupdate", (data) => {
        onTimeUpdate(data?.currentTime ?? 0, data?.duration ?? 0);
      });
    }
    ctrl.init(containerRef.current);
    controllerRef.current = ctrl;
    setController(ctrl);
    return () => {
      ctrl.destroy();
      controllerRef.current = null;
      initializedRef.current = false;
      setController(null);
    };
  }, []);
  return /* @__PURE__ */ jsxs(TanStackPlayerContext.Provider, { value: controller, children: [
    /* @__PURE__ */ jsx(
      "div",
      {
        ref: containerRef,
        className,
        style: { width: "100%", ...style },
        "data-tanstack-player": true
      }
    ),
    controller && children
  ] });
}
function normalizeSrc(src) {
  if (typeof src === "string") {
    return { src, type: inferType(src) };
  }
  return src;
}
function inferType(src) {
  if (src.endsWith(".m3u8")) return "application/x-mpegURL";
  if (src.endsWith(".mpd")) return "application/dash+xml";
  if (src.endsWith(".webm")) return "video/webm";
  if (src.endsWith(".ogg") || src.endsWith(".ogv")) return "video/ogg";
  return "video/mp4";
}
function useTanStackPlayer() {
  const controller = usePlayerContext();
  const state = useSyncExternalStore(
    useCallback(
      (onStoreChange) => controller.store.subscribe(onStoreChange),
      [controller]
    ),
    () => controller.store.getState(),
    () => controller.store.getState()
  );
  return {
    // ─── State ────────────────────────────────────
    isPlaying: state.isPlaying,
    isReady: state.isReady,
    duration: state.duration,
    currentTime: state.currentTime,
    buffered: state.buffered,
    volume: state.volume,
    playbackRate: state.playbackRate,
    isFullscreen: state.isFullscreen,
    isMuted: state.isMuted,
    error: state.error,
    // ─── Controls ─────────────────────────────────
    play: () => controller.play(),
    pause: () => controller.pause(),
    seek: (seconds) => controller.seek(seconds),
    setVolume: (value) => controller.setVolume(value),
    setMuted: (muted) => controller.setMuted(muted),
    setPlaybackRate: (rate) => controller.setPlaybackRate(rate),
    requestFullscreen: () => controller.requestFullscreen(),
    exitFullscreen: () => controller.exitFullscreen(),
    // ─── Direct controller access ─────────────────
    controller
  };
}
function useProgress() {
  const controller = usePlayerContext();
  const state = useSyncExternalStore(
    useCallback(
      (onStoreChange) => controller.store.subscribe(onStoreChange),
      [controller]
    ),
    () => controller.store.getState(),
    () => controller.store.getState()
  );
  const progress = useMemo(() => {
    if (state.duration === 0) return 0;
    return state.currentTime / state.duration * 100;
  }, [state.currentTime, state.duration]);
  return {
    currentTime: state.currentTime,
    duration: state.duration,
    buffered: state.buffered,
    /** Progress as a percentage (0–100) */
    progress
  };
}
function useBookmarks() {
  const controller = usePlayerContext();
  const [bookmarks, setBookmarks] = useState([]);
  const getPlugin = useCallback(() => {
    return controller.getPlugin("bookmark");
  }, [controller]);
  useEffect(() => {
    const refresh = () => {
      const plugin = getPlugin();
      if (plugin) {
        setBookmarks(plugin.getBookmarks());
      }
    };
    refresh();
    const unsub1 = controller.eventBus.on("bookmark:added", refresh);
    const unsub2 = controller.eventBus.on("bookmark:removed", refresh);
    return () => {
      unsub1();
      unsub2();
    };
  }, [controller, getPlugin]);
  const addBookmark = useCallback(
    (label, time) => {
      const plugin = getPlugin();
      if (!plugin) {
        console.warn("[TanStack Player] Bookmark plugin is not registered.");
        return;
      }
      return plugin.addBookmark(label, time);
    },
    [getPlugin]
  );
  const removeBookmark = useCallback(
    (id) => {
      const plugin = getPlugin();
      return plugin?.removeBookmark(id) ?? false;
    },
    [getPlugin]
  );
  const seekToBookmark = useCallback(
    (id) => {
      const plugin = getPlugin();
      return plugin?.seekToBookmark(id) ?? false;
    },
    [getPlugin]
  );
  return {
    bookmarks,
    addBookmark,
    removeBookmark,
    seekToBookmark
  };
}

export { TanStackPlayer, TanStackPlayerContext, useBookmarks, usePlayerContext, useProgress, useTanStackPlayer };
